### PR TITLE
[4.0] Implement char counter for input field, too

### PR DESCRIPTION
**Pull request by bembelimen**
### Summary of Changes
Enabled character counter for text fields

### Testing Instructions
1. Apply patch
2. NPM i
3. Open components/com_users/forms/registration.xml
4. Add the following attributes to the "name" field:
`
maxlength="30"
charcounter="true"`
5. Go to the frontend registration and check the name field

### Actual result BEFORE applying this Pull Request
Nothing different

### Expected result AFTER applying this Pull Request
grafik

grafik

### CODE REVIEW

**Comment by brianteeman** [comment]
Could this option be added to the custom fields as well?

File: libraries/src/Form/Field/TextField.php [hunk 1]
```diff
@@ -300,6 +314,7 @@ protected function getLayoutData()
 			'addonBefore' => $this->addonBefore,
 			'addonAfter'  => $this->addonAfter,
 			'options'     => $options,
+			'charcounter'  => $this->charcounter,
```
Comment (richard67) [inline libraries/src/Form/Field/TextField.php]:
```suggestion
			'charcounter' => $this->charcounter,
```

**Comment by bembelimen** [comment]
> 
> 
> Could this option be added to the custom fields as well?

In theory, it could be added to any (input/textarea) field. Currently I activate it manually in my override via:
```$form->setFieldAttribute('fieldname', 'charcounter', 'true', 'com_fields');``` and it works "out-of-the-box".

But that's currently a "bigger" problem in Joomla! (in my opinion): we use for every field an unique layout (which is good on one side) but therefore are very inconsistent over all (like the suffix/prefix info are only available for normal text fields and not for e.g. number fields etc.).
Probably worth to have there a nicer solution.

**Comment by brianteeman** [comment]
> like the suffix/prefix info are only available for normal text fields and not for e.g. number fields etc.

That is a bug that should be fixed ;)

**Comment by bembelimen** [comment]
> 
> 
> > like the suffix/prefix info are only available for normal text fields and not for e.g. number fields etc.
> 
> That is a bug that should be fixed ;)

Yes and no, it's failure by design :) [the number field does not extend the text field](https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Form/Field/NumberField.php#L26) and the [layouts](https://github.com/joomla/joomla-cms/blob/4.0-dev/layouts/joomla/form/field/number.php) could be one base layout with just different attributes....But I guess this counts for many fields (like all dropdowns etc.)

**Comment by brianteeman** [comment]
> it's failure by design :) 

or by omission

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 97294eb193b7cf0b7abef7b02a453e47a40dc1bc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by dgrammatiko** [comment]
@bembelimen I wouldn't enable that **before** fixing the counter to work nicely for repeatable fields. BTW the project should review the fields compatibility with repeatable fields and patch what's broken (ie a field with a counter in a repeatable will work only for the first instance, all others will miss the counter because there's no event to trigger the initialization...)

**Comment by bembelimen** [comment]
Hi @dgrammatiko do you mean this PR (as the problem is the same with textareas then) or the spreading over all fields?

**Comment by dgrammatiko** [comment]
> as the problem is the same with textareas

Yes, the initialization of that script is missing the listener for subform add/removal part

> the spreading over all fields?

There are many core fields that have issues or not working at all in subforms. Some are easy to patch (ie this one) others need some more changes

**Comment by brianteeman** [comment]
it will be more of an issue with this field - and I see an issue whereby the capabilities of a custom field are not in sync with an xml field.

I agree that all fields need to be reviewed with subforms - there are already a few open issues but thats beyond the scope here (I think)

File: build/build-modules-js/init/patches.es6.js [hunk 2]
```diff
@@ -27,6 +27,6 @@ module.exports.patchPackages = async (options) => {
   dest = join(mediaVendorPath, 'short-and-sweet');
   const shortandsweetPath = `${dest}/${options.settings.vendors['short-and-sweet'].js['dist/short-and-sweet.min.js']}`;
   let ShortandsweetJs = await readFile(shortandsweetPath, { encoding: 'utf8' });
-  ShortandsweetJs = ShortandsweetJs.concat('shortAndSweet(\'textarea.charcount\', {counterClassName: \'small text-muted\'});');
+  ShortandsweetJs = ShortandsweetJs.concat('shortAndSweet(\'textarea.charcount,input.charcount\', {counterClassName: \'small text-muted\'});');
```
Comment (dgrammatiko) [inline build/build-modules-js/init/patches.es6.js]:
```suggestion
  ShortandsweetJs = ShortandsweetJs.concat(`
shortAndSweet('textarea.charcount,input.charcount', {counterClassName: 'small text-muted'});
/** Repeatable */
document.addEventListener("joomla:updated", (event) => [].slice.call(event.target.querySelectorAll('textarea.charcount,input.charcount)).map((el) => shortAndSweet(el, {counterClassName: 'small text-muted'})));
`);
```

Comment (dgrammatiko) [reply build/build-modules-js/init/patches.es6.js]: It's not tested, but it should cover repeatable both for the `input` and `textarea`

**Comment by dgrammatiko** [comment]
> it will be more of an issue with this field - and I see an issue whereby the capabilities of a custom field are not in sync with an xml field.

Probably the repeatable versions of `input` and `textarea` need also an update to expose this feature (if it's not already exposed, I haven't checked the code, I was commenting on the base that the core fields should play nicely with repeatables).
Anyways I provided the code needed

File: build/build-modules-js/init/patches.es6.js (hunk 2 above)
Comment (bembelimen) [reply build/build-modules-js/init/patches.es6.js]: any reasons to go for ```textarea.charcount,input.charcount``` or should we just go for ```.charcount```?

Comment (dgrammatiko) [reply build/build-modules-js/init/patches.es6.js]: I guess the `element.class` has higher specificity that's why was implemented like that? Both will work, the `element.class` has narrower scope.

@Fedik can you check these couple of lines?

Comment (Fedik) [reply build/build-modules-js/init/patches.es6.js]: for "joomla:updated" event (eg in subform) `event.target` is the updated section, so you have to query it,
correct code is:

```javascript
document.addEventListener("joomla:updated", (event) => {
  const $els = event.target.querySelectorAll('textarea.charcount,input.charcount);
  if ($els.length) {
    shortAndSweet(Array.from($els), {counterClassName: 'small text-muted'});
  }
});
```

Comment (dgrammatiko) [reply build/build-modules-js/init/patches.es6.js]: One liner: 😉

```js
document.addEventListener("joomla:updated", (event) => [].slice.call(event.target.querySelectorAll('textarea.charcount,input.charcount)).map((el) => shortAndSweet(el, {counterClassName: 'small text-muted'})));
```

Comment (Fedik) [reply build/build-modules-js/init/patches.es6.js]: ~~the problem with oneliner here is that it unpack the array, but shortAndSweet() internally pack it back to array `[target]`~~ :wink: 

https://github.com/rikschennink/short-and-sweet/blob/20fcf8067ed9af253f914065683d28ef0fe4faba/src/short-and-sweet.js#L168

Comment (Fedik) [reply build/build-modules-js/init/patches.es6.js]: ah, no, ignore, your One liner is correct, mine is buggy :smile:

**Comment by brianteeman** [comment]
> But that's currently a "bigger" problem in Joomla! (in my opinion): we use for every field an unique layout (which is good on one side) but therefore are very inconsistent over all (like the suffix/prefix info are only available for normal text fields and not for e.g. number fields etc.).

I just checked. Suffix and Prefix are available for every field - it is in the render options

image

image

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on 97294eb193b7cf0b7abef7b02a453e47a40dc1bc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by dgrammatiko** [comment]
@richard67 my comment is still valid, introducing this without support for repeatable fields is plain wrong 😑

**Comment by richard67** [comment]
> @richard67 my comment is still valid, introducing this without support for repeatable fields is plain wrong 😑

@dgrammatiko Sorry, was not clear to me that there is something remaining to be done since the PR has got 2 test.

@bembelimen What's the status of this PR? You want to leave it as it is, or do you want to adapt it to support repeatable fields?

**Comment by richard67** [comment]
Back to pending due to ongoing clarifications.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by richard67** [comment]
Needs review.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has automatically rebased to 4.2-dev.

**Comment by mgscreativa** [comment]
Hi! just using the brand new charcounter param in my edit message form like this

		<field
			name="title"
			type="text"
			label="COM_JAPPPUSH_FIELD_MESSAGE_TITLE_LABEL"
			maxlength="35"
			charcounter="true"
			required="true"
		/>

		<field
			name="body"
			type="textarea"
			label="COM_JAPPPUSH_FIELD_MESSAGE_BODY_LABEL"
			maxlength="160"
			charcounter="true"
		/>

Is it possible to apply this to the text input?

Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by richard67** [comment]
> Is it possible to apply this to the text input?

@mgscreativa Isn't that exactly what this pull request (PR) here here is doing?

If so, please test this PR and then mark your test result by going to the issue tracker here https://issues.joomla.org/tracker/joomla-cms/35678 and then using the blue "Test this" button at the top left corner, selecting the appropriate test result and submitting.

Every PR needs 2 successful tests by a human to get accepted, so if this PR solves your problem then please test it as described to bring it forward. Thanks in advance.

**Comment by mgscreativa** [comment]
I have tested this item :white_check_mark: successfully on ae922dc3df8ed0c6a6e9f9e5a6d217342e50fcd1<br /><br />Tested ok, no issues!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by mgscreativa** [comment]
Hi @richard67 done testing, works ok! Hope it gets in J 4.1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

File: build/build-modules-js/init/patches.es6.js [hunk 3]
```diff
@@ -27,7 +27,11 @@ module.exports.patchPackages = async (options) => {
   dest = join(mediaVendorPath, 'short-and-sweet');
   const shortandsweetPath = `${dest}/${options.settings.vendors['short-and-sweet'].js['dist/short-and-sweet.min.js']}`;
   let ShortandsweetJs = await readFile(shortandsweetPath, { encoding: 'utf8' });
-  ShortandsweetJs = ShortandsweetJs.concat('shortAndSweet(\'textarea.charcount\', {counterClassName: \'small text-muted\'});');
+  ShortandsweetJs = ShortandsweetJs.concat(`
+    shortAndSweet('textarea.charcount,input.charcount', {counterClassName: 'small text-muted'});
+    /** Repeatable */
+    document.addEventListener("joomla:updated", (event) => [].slice.call(event.target.querySelectorAll('textarea.charcount,input.charcount)).map((el) => shortAndSweet(el, {counterClassName: 'small text-muted'})));
```
Comment (mgscreativa) [inline build/build-modules-js/init/patches.es6.js]: Warning! querySelectorAll('textarea.charcount,input.charcount) missing closing '! should be querySelectorAll('textarea.charcount,input.charcount')

**Review by mgscreativa — APPROVED** [review]
Tested and it works just fine!

File: build/build-modules-js/init/patches.es6.js (hunk 3 above)
Comment (richard67) [reply build/build-modules-js/init/patches.es6.js]: Hmm that seems indeed wrong, closing single quote missing on that query selector.

Comment (mgscreativa) [reply build/build-modules-js/init/patches.es6.js]: Great! thanks!

**Comment by RickR2H** [comment]
@bembelimen After building I get the error: Uncaught SyntaxError: missing ) after argument list (at short-and-sweet.min.js?1.0.4:10:104)
Don't know if it is related, but the counter does not show up. Could you take a look?

**Comment by richard67** [comment]
> @bembelimen After building I get the error: Uncaught SyntaxError: missing ) after argument list (at short-and-sweet.min.js?1.0.4:10:104)
> Don't know if it is related, but the counter does not show up. Could you take a look?

Previous review comment already mentioned that. There is indeed a syntax error in the js. The closing single quote of a query selector string is missing.

**Comment by mgscreativa** [comment]
Hi! is there anything else I can do to helpo this PR to pass? I'm using it in my site and it works just fine.

Thanks!

**Comment by richard67** [comment]
As long as @bembelimen does not fix the syntax error in the JS mentioned ant, it can not be accepted.

**Comment by mgscreativa** [comment]
Hi @richard67 can I send another PR with the same changes to try to pass it through?

**Comment by mgscreativa** [comment]
Hi @richard67 Can I send a fresh new PR to pass this new feature? Because the original PR author doesn't responds at all...

**Comment by richard67** [comment]
> Hi @richard67 Can I send a fresh new PR to pass this new feature? Because the original PR author doesn't responds at all...

@mgscreativa Let's wait a bit more. I've just notified him again.

**Comment by obuisard** [comment]
I have tested this item :white_check_mark: successfully on e13df24bd14a12237df2c4340c4631ed285eedac<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by sdwjoomla** [comment]
I have tested this item :white_check_mark: successfully on e13df24bd14a12237df2c4340c4631ed285eedac<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35678">issues.joomla.org/tracker/joomla-cms/35678</a>.</sub>

**Comment by obuisard** [comment]
Thank you Benjamin @bembelimen!

**Comment by obuisard** [comment]
Documentation found under https://docs.joomla.org/Text_form_field_type

**Pull request merged**